### PR TITLE
LLC Independence

### PR DESCRIPTION
**Pull request by xsahil03x**
## CLA

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required).
- [x] The code changes follow best practices
- [x] Code changes are tested (add some information if not applicable)

### CODE REVIEW

File: packages/dart_client/lib/src/client.dart [hunk 1]
```diff
@@ -478,15 +458,14 @@ class Client {
...
-    _connectionStatusListener = () async {
-      final value = _ws.connectionStatus.value;
-      wsConnectionStatus.value = value;
+    _connectionStatusHandler = (ConnectionStatus status) async {
+      _wsConnectionStatus = status;
       handleEvent(Event(
         type: EventType.connectionChanged,
-        online: value == ConnectionStatus.connected,
+        online: status == ConnectionStatus.connected,
       ));
```
Comment (Nash0x7E2) [inline packages/dart_client/lib/src/client.dart:466]: Missing a trailing comma here.

File: packages/dart_client/lib/src/db/chat_persistence_client.dart [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import 'package:stream_chat/src/models/channel_model.dart';
+import 'package:stream_chat/src/models/channel_state.dart';
+import 'package:stream_chat/src/models/event.dart';
+import 'package:stream_chat/src/models/member.dart';
+import 'package:stream_chat/src/models/message.dart';
+import 'package:stream_chat/src/models/reaction.dart';
+import 'package:stream_chat/src/models/read.dart';
+import 'package:stream_chat/src/models/user.dart';
+
+///
```
Comment (Nash0x7E2) [inline packages/dart_client/lib/src/db/chat_persistence_client.dart:11]: Can you add a brief description of the class? It does not have to be in-depth, maybe just an overview of its function and responsibility.

File: packages/dart_client/lib/src/db/chat_persistence_client.dart [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+  /// Remove a channel by cid
+  Future<void> deleteChannels(List<String> cids);
+
+  /// Update messages data from a list
+  Future<void> updateMessages(String cid, List<Message> messages);
+
+  /// Get the info about channel threads
+  Future<Map<String, List<Message>>> getChannelThreads(String cid);
+
+  ///
```
Comment (Nash0x7E2) [inline packages/dart_client/lib/src/db/chat_persistence_client.dart:117]: Can you add a simple one-liner to these functions? We can work on this together if you like 😄

File: packages/dart_client/lib/src/db/chat_persistence_client.dart [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    final deleteReactions = deleteReactionsByMessageId(channelStates
+        .expand((it) => it.messages)
+        .map((m) => m.id)
+        .toList(growable: false));
+
+    final deleteMembers = deleteMembersByCids(
+      channelStates.map((it) => it.channel.cid).toList(growable: false),
+    );
+
+    await Future.wait([
```
Comment (Nash0x7E2) [inline packages/dart_client/lib/src/db/chat_persistence_client.dart:199]: This is very cool, great job!

File: packages/flutter_widgets/example/lib/notifications_service.dart [hunk 5]
```diff
@@ -34,15 +37,24 @@ void showLocalNotification(Message message, ChannelModel channel) async {
...
-  final messageId = notification['data']['id'];
-
-  final notificationData =
-      await NotificationService.getAndStoreMessage(messageId);
-
-  showLocalNotification(
-    notificationData.message,
-    notificationData.channel,
-  );
+  // final messageId = notification['data']['id'];
```
Comment (Nash0x7E2) [inline packages/flutter_widgets/example/lib/notifications_service.dart:40]: If this code is not used, it should be removed.

File: packages/flutter_widgets/lib/src/connection_status_builder.dart [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import 'package:flutter/material.dart';
+import 'package:stream_chat_flutter_core/stream_chat_flutter_core.dart';
+import 'stream_chat.dart';
+
+class ConnectionStatusBuilder extends StatelessWidget {
```
Comment (Nash0x7E2) [inline packages/flutter_widgets/lib/src/connection_status_builder.dart:5]: Consider adding a brief description of the class.

File: packages/stream_chat_flutter_core/lib/src/stream_chat_core.dart [hunk 7]
```diff
@@ -4,6 +4,8 @@ import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
 import 'package:stream_chat/stream_chat.dart';
 
+typedef EventHandler = void Function(Event event);
```
Comment (Nash0x7E2) [inline packages/stream_chat_flutter_core/lib/src/stream_chat_core.dart:7]: This should be moved to `typedefs.dart`

File: packages/stream_chat_flutter_core/pubspec.yaml [hunk 8]
```diff
@@ -10,8 +10,8 @@ environment:
   flutter: ">=1.17.0"
 
 dependencies:
-  stream_chat: ^0.2.23+3
-  rxdart: ^0.24.1
+  stream_chat:
+    path: ../dart_client
```
Comment (Nash0x7E2) [inline packages/stream_chat_flutter_core/pubspec.yaml:14]: Consider using the version number instead, melos boostrap automatically handles linking.

File: packages/stream_chat_persistence/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+# stream_chat_persistence
+
+A new Flutter package.
```
Comment (Nash0x7E2) [inline packages/stream_chat_persistence/README.md:3]: I think we should update this with a short description of the package and its intended use. 

@imtoori and I can help with this.

File: packages/stream_chat_persistence/lib/src/db/shared/native_db.dart [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+import 'dart:isolate';
+import 'package:moor/ffi.dart';
+import 'package:moor/isolate.dart';
+import 'package:moor/moor.dart';
+import 'package:path/path.dart';
+import 'package:path_provider/path_provider.dart';
+
+import '../moor_chat_database.dart';
+
+class SharedDB {
```
Comment (Nash0x7E2) [inline packages/stream_chat_persistence/lib/src/db/shared/native_db.dart:13]: Please add doc comments to this class. Omitting doc comments for public apis will result in a low pub.dev score.

File: packages/stream_chat_persistence/lib/src/stream_chat_persistence_client.dart [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        return;
+      case ConnectionMode.background:
+        _logger.info('Connecting on background isolate');
+        _db = await SharedDB.constructOfflineStorage(userId);
+        return;
+    }
+  }
+
+  @override
+  Future<Event> getConnectionInfo() {
```
Comment (Nash0x7E2) [inline packages/stream_chat_persistence/lib/src/stream_chat_persistence_client.dart:52]: Consider adding doc comments to these functions.

File: packages/stream_chat_persistence/pubspec.yaml [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+
+environment:
+  sdk: ">=2.7.0 <3.0.0"
+
+dependencies:
+  moor: ^3.4.0
+  path: ^1.7.0
+  path_provider: ^1.6.27
+  sqlite3_flutter_libs: ^0.3.0
+  stream_chat:
```
Comment (Nash0x7E2) [inline packages/stream_chat_persistence/pubspec.yaml:16]: Consider using the current version defined in `stream_chat`. Melos will automatically link.

**Review by Nash0x7E2 — CHANGES_REQUESTED** [review]
Fantastic job on this PR! Your code is succinct and very easy to follow! 

There are a few areas I think we can improve the documentation and styling. Please see the review comments.

**Comment by Nash0x7E2** [comment]
@xsahil03x Looks like the analyzer found a few issues:

```bash
[example]: 
[example]:   info - Unused import: 'dart:io' at lib/advanced_options_page.dart:1:8 - (unused_import)
[example]:   info - Unused import: 'package:example/routes/app_routes.dart' at lib/advanced_options_page.dart:3:8 - (unused_import)
[example]:   info - Unused import: 'package:stream_chat_persistence/stream_chat_persistence.dart' at lib/advanced_options_page.dart:11:8 - (unused_import)
[example]:   info - Unused import: 'package:streaming_shared_preferences/streaming_shared_preferences.dart' at lib/advanced_options_page.dart:12:8 - (unused_import)
[example]:   info - Unused import: 'dart:io' at lib/main.dart:2:8 - (unused_import)
[example]:   info - Unused import: 'package:example/main.dart' at lib/notifications_service.dart:3:8 - (unused_import)
[example]:   info - Unused import: 'package:firebase_messaging/firebase_messaging.dart' at lib/notifications_service.dart:4:8 - (unused_import)
[example]: 
[example]: 7 issues found; analyzed 28 source files in 54.5s.
[stream_chat_flutter_core]: Checking project stream_chat_flutter_core...
[stream_chat]: 
[stream_chat]:   info - Unused import: 'package:stream_chat/src/models/channel_model.dart' at lib/src/client.dart:11:8 - (unused_import)
[stream_chat]: 
[stream_chat]: 1 issue found; analyzed 37 source files in 19.0s.
[stream_chat_persistence]: Checking project stream_chat_persistence...
[stream_chat_flutter]: 
[stream_chat_flutter]:   info - Unused import: 'dart:io' at example/lib/advanced_options_page.dart:1:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'package:example/routes/app_routes.dart' at example/lib/advanced_options_page.dart:3:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'package:stream_chat_persistence/stream_chat_persistence.dart' at example/lib/advanced_options_page.dart:11:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'package:streaming_shared_preferences/streaming_shared_preferences.dart' at example/lib/advanced_options_page.dart:12:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'dart:io' at example/lib/main.dart:2:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'package:example/main.dart' at example/lib/notifications_service.dart:3:8 - (unused_import)
[stream_chat_flutter]:   info - Unused import: 'package:firebase_messaging/firebase_messaging.dart' at example/lib/notifications_service.dart:4:8 - (unused_import)
[stream_chat_flutter]: 
```

File: packages/stream_chat_persistence/lib/src/stream_chat_persistence_client.dart (hunk 11 above)
Comment (xsahil03x) [reply packages/stream_chat_persistence/lib/src/stream_chat_persistence_client.dart:52]: Overridden function and member copies the doc comments from the parent class by default, So I don't think it's necessary here.

File: packages/dart_client/lib/src/db/chat_persistence_client.dart (hunk 4 above)
Comment (xsahil03x) [reply packages/dart_client/lib/src/db/chat_persistence_client.dart:210]: Thank you.

**Review by Nash0x7E2 — APPROVED** [review]
LGTM.

**Pull request merged**